[PF-1992] Migration orchestrator (#4946)
toptal/picasso: docs/modernization/picasso-code-review-rules.md

> # Picasso code review rules

Author: Review this document

Author: There are some conflicts also:

*  Inline style framing - needs reframing and different priority 
  * Idea is to use reasoning about why inline styles are needed (can we simplify DOM, does it look geometrical and visually correct) and to first exhaust base-ui customization mechanisms, also to ask for guidance in case of last solution 
*  Slot-based styling section - only legacy and specific for Tier3.b components not new pattern
*  Design patterns R5 and R10 are in confict with existing codebase - be sure for migration we keep as is and leave for later refactor and for migration to not introduce new ones

* Remove patch rules after master rebase
* Resolving of versionBump is in conflict between manifest and rules

### Source documents synthesized here

- `PICASSO_COMPONENT_DESIGN_PATTERNS.md` (repo root) — props/API spec, 16 + 3 rules
- `docs/migration/references/code-standards.md` — file structure, types, JSDoc, Tailwind composition, ESLint, tests, changesets, CI
- `docs/migration/references/practices.md` — graduated migration patterns
- `docs/migration/references/base-ui-styling.md` — Base UI v1 styling doctrine
- `docs/migration/references/design-patterns-addendum.md` — existing-violations carve-out + architectural exceptions
- `docs/migration/references/visual-verification.md` — Playwright + Happo workflow (reviewer-relevant subset distilled here)
- `docs/migration/references/happo-iteration.md` — Happo classification matrix
- `docs/migration/rules/styling.md` — non-negotiable Tailwind/Base UI styling rules
- `docs/migration/rules/package-and-build.md` — pnpm / lockfile / build policy
- `docs/migration/rules/jss-to-tailwind-crib.md` — JSS → Tailwind pattern table + worked examples
- `docs/contribution/{component-api,unit-testing,changeset-guidelines,github-workflow,visual-testing,accessibility,packages-architecture}.md`

### Out of scope (intentionally excluded)

> - Orchestrator-internal code rules (`bin/lib/*.ts` ESLint rules — see `CLAUDE.md` §"Code style for orchestrator")

Reviewer: I think way more files are outside of those conventions and rules 🤔

Author: It collected stuff from patterns, practices, references, standards and rules. This document will not be used, it's just for purpose of review and changes will be pushed to origin documents.

### R4 — Mirror native HTML prop names · **RULE**

When a prop corresponds to a native HTML attribute, use the native name verbatim: `name`, `value`, `type`, `disabled`, `placeholder`, `checked`, `readOnly`, `autoFocus`, `href`, `target`, `rel`, `alt`, `src`. Do not rename, prefix, or alias (no `inputName`, `fieldValue`, `isDisabled`, `linkHref`). For event handlers, keep the native name (`onChange`, `onBlur`, `onFocus`, `onClick`).

**Callback signatures may diverge**: when only a derived value is useful, accept the value directly rather than re-passing the event:

```tsx
// Acceptable — simpler shape, same native name
onChange?: (value: string) => void
```

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §4)

### R5 — Style overrides only via `className` or `style` · **RULE**

Consumers may customize a component's appearance exclusively through the `className` and `style` props. Do not expose other styling hooks: `classes`, `styles`, `sx`, `css`, theme overrides, slot-level class props, or styled-component wrappers.

**Why**: every additional styling hook becomes a maintenance contract. `className` + `style` is sufficient when paired with `twMerge` (consumer-last wins).

> **Migration exceptions** (see Appendix A): Modal, Typography, Dropdown, OutlinedInput temporarily retain narrowed `classes?: { ... }` shapes per audit-backed Tier 3.b decisions.

Reviewer: nit: there should be no exceptions ideally, for the migration probably we have to make them

Author: Yes, this is temp things because there is current technical debt that violates it and migration will keep debt for sake of backward compatibiltiy. I can ad a bit more to emphasize it's temporary.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §7)

### R8 — Align token names with the BASE design system · **RULE**

Names for colors, typography, sizes, spacings, and other design tokens must match those defined in the BASE design system. Do not introduce local synonyms or renamed aliases.

**Why**: design, code, and documentation share a single vocabulary. Local aliases create drift between Figma, the design-tokens package, and the runtime.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §8; tokens listed in `docs/migration/tokens/picasso-tailwind-tokens.md`)

### R9 — Use `variant` for visual variations · **RULE**

When a component has multiple visual styles, expose them through a single `variant` prop typed as a string-literal union — never split across multiple boolean flags or ad-hoc prop names.

```tsx
// Good
variant?: 'rectangle' | 'circular'

// Bad — booleans multiply combinatorially
isRectangle?: boolean
isCircular?: boolean
isOval?: boolean
```

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §9)

### R10 — Extend `BaseProps` · **RULE**

> Every component's Props interface must extend `BaseProps`, which provides the shared root-element contract: `className?: string`, `style?: CSSProperties`, `'data-testid'?: string`. Do not redeclare these props locally.

Reviewer: ```suggestion
Every component's Props interface must extend `BaseProps`, which provides the shared root-element contract: `className?: string`, `style?: CSSProperties`, `'data-testid'?: string`.
```
nit: I think this is extra 🤔

Author: Ok, this is R10 from PICASSO_COMPONENT_DESIGN_PATTERNS.md

```ts
import type { BaseProps } from '@toptal/picasso-shared'

interface Props extends BaseProps {
  /** ... */
  variant?: 'primary' | 'secondary'
}
```

> **Migration exception** (see Appendix A): 20/28 components currently extend `BaseProps`; 3/28 still extend `StandardProps`; 5/28 mixed. **Do NOT preemptively rebuild prop interfaces** mid-migration — that's a separate refactor track.

Reviewer: 👍

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §10)

### R11 — Use `as` to change the rendered element · **RULE**

When a component needs to render as a different HTML tag, expose this through an `as` prop. Prefer narrowing to specific tags the component supports rather than accepting any element type. Do not introduce custom alternatives like `tag`, `component`, or `element`.

```ts
// Preferred — narrow to what the component actually supports
as?: 'a' | 'button'

// Acceptable when truly polymorphic
as?: React.ElementType<React.HTMLAttributes<HTMLElement>>
```

**Internal translation**: Picasso's external `as` API is translated to Base UI's `render` mechanism internally — see §5 "Render prop". For button-default Base UI parts (Button, Menu.Trigger, Tabs.Tab, NumberField.Increment/Decrement, Toolbar.Button), the tag swap MUST pair with `nativeButton={false}`.

> **No runtime guards**: do NOT add `typeof`/`isValidAs` checks for `as` — TypeScript constrains it; reviewers will ask for removal.

Reviewer: nit: I think this has to be a global rule of some sort, maybe in Claude.md

Author: Agree, will mvoe from code-standards.md to PICASSO_COMPONENT_DESIGN_PATTERNS.md

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §11; `code-standards.md §"prop-by-prop boundary"`)

### R12 — Use the shared size scale · **RULE**

Expose size as a `size` prop typed against the shared `SizeType` helper, picking the subset the component supports:

```ts
import type { SizeType } from '@toptal/picasso-shared'

size?: SizeType<'small' | 'medium' | 'large'>
```

Canonical scale (from `@toptal/picasso-shared`):

```ts
export type Sizes =
  | 'xxsmall' | 'xsmall' | 'small'
  | 'medium' | 'large' | 'xlarge'

export type SizeType<T extends Sizes> = T
```

> Do NOT introduce custom size names (`tiny`, `big`, `huge`) or numeric scales.

Reviewer: nit: I think it's important to explain here, that those Size values from the type use the same language and names as BASE design system

Author: PICASSO_COMPONENT_DESIGN_PATTERNS.md §12 (R12) + §8 (R8 BASE alignment)

Agree. Also check cross-reference.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §12)

### R13 — Use the shared color palette and shade scale · **RULE**

Color props must draw base color names from the canonical `Palette` and shade names from `ColorSample`, exposing only the subset the component supports.

```ts
interface ColorSample {
  lightest?: string
  lighter?: string
  lighter2?: string
  light?: string
  light2?: string
  main?: string
  main2?: string
  dark?: string
  darker?: string
}

interface Palette {
  blue: SimplePaletteColorOptions
  green: SimplePaletteColorOptions
  yellow: SimplePaletteColorOptions
  red: SimplePaletteColorOptions
  purple: SimplePaletteColorOptions
  gradients: {
    blue: string
    darkerBlue: string
    lightGrey: string
    grey: string
    darkerGrey: string
  }
}
```

> Do NOT invent new color or shade names (`bright`, `pale`, `accent`, `orange`) or use raw hex/rgb values in the public API.

Reviewer: ```suggestion
Do NOT invent new color or shade names (`bright`, `pale`, `accent`, `orange`), do NOT use raw hex/rgb values in the public API.
```

Author: PICASSO_COMPONENT_DESIGN_PATTERNS.md §13

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §13)

### R14 — No `is` prefix on boolean props · **RULE**

Boolean props are named with the adjective only: `open`, `disabled`, `loading`, `selected`, `collapsed`, `expanded`, `active`, `hovered`, `indeterminate`. NEVER `isOpen`, `isDisabled`, `isLoading`. The same applies to `has`/`should` prefixes.

> **Why**: 28/28 current compliance — Picasso's most consistent rule. New code must match.

Reviewer: ```suggestion
```

Author: Collection-doc commentary derived from _survey-findings.md

it's actually not rule, just survey data. Ignore.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §14)

### R15 — Use compound components for multi-part components · **RULE**

When a component has distinct, composable sub-parts, expose them as static properties on the parent (`Modal.Title`, `Modal.Content`, `Modal.Actions`) rather than as separate named exports or as content props. Consumers compose the parts as children:

```tsx
<Modal open={open} onClose={onClose}>
  <Modal.Title>Title</Modal.Title>
  <Modal.Content>…</Modal.Content>
  <Modal.Actions>
    <Button onClick={onClose}>Cancel</Button>
  </Modal.Actions>
</Modal>
```

Apply only when the design genuinely requires a compound consumer API (3+ distinct sub-parts the consumer must compose). For one or two parts, keep the component monolithic.

Current adopters: Modal, Accordion, Drawer, Tabs, Menu, Dropdown, Note (plus the Button family internally). See §3 for the canonical `<Component>Compound/index.ts` wrapper-file pattern.

> 

Reviewer: for the migration we just need to ignore this rule, as we don't plan to change API of components, regardless if this pattern used now correctly or not

Author: Agree, it's already doing it through api-preservation in rules of migration. But we can be more explicit.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §F2)

### F3 — Render through `PicassoField` (or a descendant) · **RULE**

Internally, every form component must use `PicassoField` — or a wrapper that itself builds on `PicassoField` (e.g., `OutlinedInput`, `InputBase`-style descendants) — to render its field chrome. This centralizes label, hint, error, required-marker, and layout behavior.

Do NOT reimplement field chrome ad hoc.

(source: `PICASSO_COMPONENT_DESIGN_PATTERNS.md` §F3)

---

## 3. Code organization & file structure

### Canonical layout · **RULE** (22/28 conform)

```
packages/base/<Component>/src/<Component>/
├── index.ts          # public exports (named + default + types)
├── <Component>.tsx   # implementation + Props interface
├── styles.ts         # Tailwind class-building functions (pure, return string[])
├── test.tsx          # unit tests
├── story/            # *.example.tsx files (PicassoBook API)
│   └── *.example.tsx
└── __snapshots__/    # generated
```

> Variance (6/28): components with logical sub-components co-locate them in the same dir (Button has `ButtonBase.tsx`, `ButtonCheckbox.tsx` — not split into sibling dirs).

Reviewer: ```suggestion
If it makes sense to have sub-components co-locate them in the same dir (ex. Button has `ButtonBase.tsx`, `ButtonCheckbox.tsx` — not split into sibling dirs).
```

**Hooks and utilities**: when a component needs custom hooks or helpers, co-locate as `hooks.ts` or `<use-hook-name>.ts` inside the component folder — never split into a parallel `packages/.../hooks/` directory.

(source: `code-standards.md §"Component file structure"`)

### Compound-component wrapper pattern · **Convention** (6/28 use it)

When a component exposes a `Component.Item` / `Component.Tab` / `Component.Title`-style compound API, Picasso splits across two files:

- `<Component>.tsx` — the main functional component, unchanged shape.

> - `<Component>Compound/index.ts` — wraps the main export and attaches static properties (`Tabs.Tab = TabsTab`, `Menu.Item = MenuItem`).

Reviewer: ```suggestion
- `<Component>Compound/index.ts` — wraps the main export and attaches static properties (`Tabs.Tab = TabsTab`, `Menu.Item = MenuItem`).

For example:
\```
import { FormLabel } from '@toptal/picasso-form-label'

import { Form } from '../Form'
import { FormError } from '../FormError'
import { FormField } from '../FormField'
import { FormHint } from '../FormHint'
import { FormWarning } from '../FormWarning'

export const FormCompound = Object.assign(Form, {
  Field: FormField,
  Hint: FormHint,
  Error: FormError,
  Label: FormLabel,
  Warning: FormWarning,
})
\```

```

Author: docs/migration/references/code-standards.md §"Compound-component wrapper pattern"

Tests import `TabsCompound as Tabs` to access the compound API. Examples:

- `Tabs.tsx:1-128` + `TabsCompound/index.ts` exposes `Tabs.Tab`.
- `Menu.tsx:23-118` + `MenuCompound/index.ts` exposes `Menu.Item`.
- `Dropdown.tsx:125-337` + `DropdownCompound/index.ts` exposes `Dropdown.Arrow`.
- `Note.tsx:9-37` + `NoteCompound/index.ts` exposes `Note.Title`, `Note.Subtitle`, `Note.Content`.

> Apply this split only when the design genuinely requires a compound consumer API (3+ distinct sub-parts). Don't introduce a `XCompound/` wrapper just to attach a single `.X.Item` — keep simple components monolithic.

Reviewer: ```suggestion
```

Author: Agree, duplication, R15 in design-patterns defines it already.

// In the body:
<DropdownContext.Provider value={{ close: forceClose }}>
  {children}
</DropdownContext.Provider>
```

Children call `useDropdownContext()?.close()`. This is the canonical Picasso pattern for compound-internal state sharing — do NOT thread callbacks through props or use refs for this purpose.

(source: `code-standards.md §"Context-based coordination for compound parts"`)

### `PrivateProps` / `PublicProps` split · **Convention** (1/28 — Notification)

When a component has internal-only props (e.g., `elevated`, `icon` consumed by internal composition but not part of the public API), use the split:

```ts
// Notification.tsx:19-40
interface PrivateProps extends StandardProps {
  /** Internal: drawn with elevation when nested inside Modal */
  elevated?: boolean
  // ... other internal props
  /** Public visible props */
  variant?: 'success' | 'info' | 'warning'
}

export type PublicProps = Omit<PrivateProps, 'elevated' | 'icon'>
// Component takes PrivateProps internally
// index.ts exports only PublicProps as the public type
```

This pattern is preferred over commenting "// internal use only" — the type system enforces the API boundary.

> (source: `code-standards.md §"PrivateProps / PublicProps split"`)

Reviewer: ```suggestion
```

I'd suggest to remove this rule, as it might confuse LLM more than help 😃 those are pretty unique situations, as I remember

Author: Agree, it's only 1 case of it, not strong pattern.

### Export & component conventions · **RULE** (26/28 conform)

- **Named export PREFERRED, default export ALSO provided.**
- `forwardRef` wraps a **NAMED inner function** (not anonymous) — sets `displayName` correctly without a separate assignment.
- Set `Component.displayName = 'Component'` explicitly anyway (some sub-components need it for DevTools after compound attachment).
- `index.ts` re-exports the default + named, and re-exports the Props type as `<Component>Props`:

```tsx
// Component.tsx
export const Button = forwardRef<HTMLButtonElement, Props>(function Button(props, ref) {
  // ...
})
Button.displayName = 'Button'
export default Button
```

```ts
// index.ts
export { default } from './Button'
export { Button } from './Button'
export type { Props as ButtonProps } from './Button'
```

(source: `code-standards.md §"Export & component conventions"`)

### Props interface declaration · **RULE** (28/28 use `interface`)

- Use `interface Props extends ...`, NEVER `type Props = { ... }`.

> - For boolean props: bare adjective only — `disabled`, `checked`, `loading`, `selected`, `collapsed`, `open`, `expanded`, `active`, `hovered`, `indeterminate` (R14).

Reviewer: I think this rule already covered in design patterns

Author: Agree.

- **Default values via destructuring** in the function signature, NEVER `Component.defaultProps = { ... }` static assignment (legacy anti-pattern; 26/28 components avoid it).

```tsx
forwardRef<HTMLButtonElement, Props>(function Button(
  { disabled = false, variant = 'primary', size = 'medium', ...rest },
  ref
) { /* ... */ })
```

> **Carve-out**: `Dropdown.tsx` exposes `defaultProps?: Partial<PropsWithBaseSpacing>` in its Props type for overload support — that's a type-level field name, not the runtime anti-pattern. Don't introduce a new `defaultProps` type field unless you genuinely need overload support.

Reviewer: looks like a very unique exception, that I'd not be showing as rules

Author: Agree.

- **Required** on EVERY public Props interface field.
- Format: single-line `/** description */` immediately above the property.

```ts
interface Props extends BaseProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
}
```

- **Forbidden** on passthrough internal props (`ownerState`, MUI-Base injected props, `data-private`, `data-testid` from BaseProps). These surface in TS doc generation as public API.
- For `@deprecated` JSDoc tags: should include a Jira reference `[ABC-1234]` or URL. ESLint `todo-plz/ticket-ref` is configured as **warn** (not error), so non-compliant `@deprecated` tags ship without blocking CI — reviewers consistently flag this regardless.

(source: `code-standards.md §"JSDoc rules"`)

### Reserved props · **Convention**

These props are framework-reserved; pass through without surfacing in JSDoc:

- `data-private?: string` — framework hook (theme access / analytics). Used internally by Tooltip, Slider, Switch, Checkbox.
- `data-testid?: string` — from `BaseProps`; never declare per-component.

> - `ownerState`, MUI-Base injected props — implementation detail.

Reviewer: it has been used once and it already to put to rules 😃

Author: This file is collection of all context, I named it rules, but actually it's rules, references, ... I think I need to add source of each "rule" so you can understand level of importance (hard, medium, ...)

Author: It's used more, data-private in 4 components and `ownerState` in any @mui/base component. But i think we can structure it differently in JSDoc rules under docs/migration/references/code-standards.md

If your migration touches a component that uses these, preserve the pass-through; do NOT add JSDoc to them.

(source: `code-standards.md §"Reserved props"`)

### Import conventions · **RULE**

**Order** (observed in 28/28 components surveyed):

1. React / react-related (`react`, `react-dom`)
2. Third-party (`@mui/base`, `@material-ui/*`, `@base-ui/react`, etc.)
3. `@toptal/*` packages

> 4. Relative imports (`./styles`, `../utils`)

Reviewer: I think this is already enforced by eslint autofixable rules, if I'm not mistaken

Author: Will check what is source of this rule because some rules are actually extracted from eslint and this could actually be one.

> **Barrel imports preferred**: `import { StandardProps, SizeType } from '@toptal/picasso-shared'` (not deep paths).

Reviewer: ```suggestion
**Barrel imports preferred**: `import { StandardProps, SizeType } from '@toptal/picasso-shared'` (NEVER deep paths).
```

> **Forbidden**: import from `@toptal/picasso` within sub-packages (ESLint error `@toptal/davinci/no-package-self-imports`). Each sub-package imports directly from sibling packages, never via the aggregate.

Reviewer: I think this deserves a separate rule maybe or eslint rule/error if it does not exist yet

Author: Will check what is source of this rule because some rules are actually extracted from eslint and this could actually be one.

Author: It's eslint error already.

#### Where to import what

| Symbol | Source |
|---|---|
| `StandardProps`, `BaseProps`, `TextLabelProps`, `SizeType`, `OverridableComponent`, `useIsomorphicLayoutEffect` | `@toptal/picasso-shared` |
| `render` (test wrapper), `fireEvent`, test types | `@toptal/picasso-test-utils` |
| `noop`, `usePageScrollLock`, ref utilities | `@toptal/picasso-utils` |
| `twMerge`, `twJoin` | `@toptal/picasso-tailwind-merge` |
| `cx` | `classnames` |

`withClasses` from `@toptal/picasso-utils` is **deprecated** — do not introduce new usages.

(source: `code-standards.md §"Import conventions"`)

### Custom hooks · **Convention** (4+ examples)

- `use*` prefix universal.

> - Co-locate inside the component folder (`Slider/hooks.ts`, `Tooltip/use-tooltip-state.ts`).

Reviewer: seems to me that it's better to enforce one of those options

Author: Codebase have both options, do you have preference? I like use-{hook}.ts more. What about you?

- Examples: `useLabelOverlap` (Slider), `useTooltipState` (Tooltip), `useOnScreen` (Slider), `useCombinedRefs` (shared utility for merging user + internal refs).

(source: `code-standards.md §"Custom hooks"`)

### Re-export pattern · **RULE** (universal)

The package's top-level `src/index.ts` re-exports every component and its Props type:

```ts
export { default as Button, type ButtonProps } from './Button'
export { default as Modal, type ModalProps } from './Modal'
```

> **Do NOT change this shape during a migration** — the package's public API surface is contractual.

Reviewer: everything that is exported in index.ts has to be also exported from `@toptal/picasso` `src/index.ts`

Author: New rule 👍

(source: `code-standards.md §"Re-export pattern"`)

### `*.example.tsx` story files · **Convention** (universal)

- Location: `<Component>/story/<Domain>.example.tsx`. NEVER a central `stories.ts` index.
- One example per file. Filename in PascalCase, matches the story title kebab-cased in URLs.

> - Story registration via `addExample` calls in a `story/index.jsx` (one per component) — read this file to discover the component's full story list.

Reviewer: I think it might be more efficient if we give an example of Example here, because there are more functions from PicassoBook available rather than `addExample` only

Author: This should reference docs/migration/references/code-standards.md and in there we can add proper example if needed.

- `.example.tsx` files exempt from SSR rules, multi-comp rule, and a number of `@typescript-eslint` rules.

(source: `code-standards.md §"*.example.tsx story files"`)

---

## 4. Types & casting

Zero violations of these rules across the 28 migration-scope components today. Strict enforcement.

### No `any` in component source · **RULE** (ESLint `error`)

ESLint `@typescript-eslint/no-explicit-any` is **error** in source, off in tests. Tests may cast mock objects.

**Out-of-scope legacy**: 2 pre-existing `any` violations exist in untouched code (`TreeViewContainer.tsx:43`, `List/utils/generateListItems.tsx:6`) and 2 legacy `// @ts-ignore` comments in `Dropdown.tsx:223,227` with TODO+Jira markers. Those are explicit refactor markers, NOT precedents to extend.

(source: `code-standards.md §"Type-narrowing & casting"`)

> ### No `as unknown as T` blanket casts · **RULE**

Reviewer: ```suggestion
### No types casting · **RULE**
```

Author: It's a bit too generic and restricting imo, check Type alignment at the boundary sections:

https://github.com/toptal/picasso/blob/feature/pf-1992-migration-orchestrator/docs/migration/references/code-standards.md#the-prop-by-prop-boundary--canonical-resolution-for-root-element-type-mismatches

https://github.com/toptal/picasso/blob/889b67481e97f7d59d5cf3ba457e6d6707a2ee8c/docs/migration/rules/base-ui-react-api-crib.md#type-alignment-at-the-boundary-dont-weaken-the-public-api

In component source. Tests may cast mock objects.

(source: `code-standards.md §"Type-narrowing & casting"`)

### No bare `// @ts-ignore` · **RULE**

If you absolutely need to suppress a type, use `@ts-expect-error <reason>` so the suppression is self-documenting and fails the build when the underlying type drifts.

(source: `code-standards.md §"Type-narrowing & casting"`)

### The "prop-by-prop boundary" — canonical resolution for root-element-type mismatches · **RULE**

When `@base-ui/react`'s root part has a type that doesn't fully line up with your public `Props` (e.g. `Props extends ButtonHTMLAttributes<HTMLButtonElement>` but `BaseUISwitch.Root.Props` doesn't extend `ButtonHTMLAttributes`), the canonical fix is **NOT**:

> #### Anti-pattern A — blanket cast

Reviewer: it seems antipatterns have more examples and context space than patterns of how it should do. I'd suggest either to remove those either add more examples of how it is correct to do. LLM might be confused

Author: Agree, problem is real but we need positive example and reduce anti-pattern duplication

// shape are destructured out (or transformed); everything else spreads
// through unchanged. Public API parity preserved, no cast, no allowlist.
const Switch = (props: Props) => {
  const {
    onChange,     // signature differs — adapt to onCheckedChange below
    checked,      // base-ui uses checked: boolean directly, but we want to clamp
    ...rest       // ← all other ButtonHTMLAttributes flow through
  } = props
  return (
    <BaseUISwitch.Root
      {...rest}
      checked={checked ?? false}
      onCheckedChange={c => onChange?.(syntheticEvent(c), c)}
    />
  )
}
```

**How to find which props to destructure**: open `node_modules/@base-ui/react/<group>/<part>/<Part>.d.ts` and diff its `*.Props` interface against your public `Props`. The intersection's NAME-OVERLAPS-WITH-DIFFERENT-TYPES set is what you destructure. Everything else is type-compatible and spreads.

**Empirical sanity check**: for Tier 0 components the destructure list is typically 1–3 props (onChange signature, occasional value/checked clamp, sometimes a removed prop). If you find yourself destructuring 6+ props, re-read the library's `.d.ts` — you're sliding into Anti-pattern B.

> The migration-period oscillation observed on Switch review-iter 7 (2026-05-22) — allowlist → cast → allowlist across three consecutive iters — was the agent flipping between the two anti-patterns above because neither it nor the audit-agent named the canonical third option. Cite this section directly in PR replies when reviewers raise the question.

Reviewer: it seems those rules more confusing than helpful 🤔

Author: I think they are extracted from lesson learned, I will ask agent how useful it is and can it be more generalized.

(source: `code-standards.md §"prop-by-prop boundary"`, `practices.md §"API preservation"`)

### Type alignment at the boundary · **RULE**

Cast at the type BOUNDARY (helper return type, local typed binding), NOT at the JSX call site:

```tsx
// Preferred — hoist cast into helper return type:
const getClickHandler = (
  loading?: boolean,
  handler?: Props['onClick']
): BaseUIButton.Props['onClick'] =>
  (loading ? noop : handler) as BaseUIButton.Props['onClick']

// Then JSX is clean:
<BaseUIButton onClick={getClickHandler(loading, onClick)} />
```

This isolates the type-system compromise to one location instead of sprinkling casts across the render. Reviewers can verify the boundary cast in one place.

(source: `code-standards.md §"prop-by-prop boundary"`, `rules/base-ui-react-api-crib.md §"Type alignment at the boundary"`)

### Polymorphic + ref forwarding · **RULE**

- `forwardRef<HTMLButtonElement, Props>(...)` already types `ref` correctly. Don't cast `ref` at the JSX site.

> - Spreading `{...rest}` with a cast (`{...(rest as BaseUIButton.Props)}`) is `// @ts-ignore` in disguise. If `rest` doesn't conform, drop the offending Picasso-only prop BEFORE spreading. NEVER fall back to `any`.

Reviewer: ```suggestion
```

Author: Why, can you explain?

Author: I think we need to give positive example of it. Something like "Drop Picasso-only props from rest before spreading to a @base-ui/react part."

| **3. `data-*` state attributes** | State-driven styling without React subscriptions | Hover, open, checked, disabled, side-positioning, animation phases |
| **4. CSS variables (`--var`)** | Values Base UI computes (positions, sizes, transform origins) | Position-anchored animation, popup sizing, geometry-driven layout |
| **5. `style` prop (static or function-of-state)** | Inline styles | Last resort — computed values that cannot be expressed as classes |

Every part exposes the consistent signature:

```ts
className?: string | ((state: State) => string | undefined);
style?:     React.CSSProperties | ((state: State) => React.CSSProperties | undefined);
render?:    ReactElement | ((props: HTMLProps, state: State) => ReactElement);
```

(source: `base-ui-styling.md §2`)

### `className` composition — `twMerge(cx(...))` · **RULE**

Every Tailwind-on-headless codebase needs a class-merging pipeline that resolves conflicts. Picasso uses `classnames` (`cx`) + `twMerge` from `@toptal/picasso-tailwind-merge`:

```ts
import cx from 'classnames'
import { twMerge } from '@toptal/picasso-tailwind-merge'

> twMerge(cx('px-4 text-sm', isLarge && 'px-6 text-base'), className)

Reviewer: I think usage of `cx` is not needed with twMerge, example:

```
twMerge(
                'relative min-h-0 flex overflow-hidden',
                classesByOrientation[orientation].root,
                classesByVariant[variant].root,
                className
              )
```

https://github.com/toptal/picasso/blob/fc2c4e762dc986c54da278457813ba5cbb7e30b4/packages/base/Tabs/src/Tabs/Tabs.tsx#L98-L103

Author: Agree.

Author: Note for agent: 

No cx needed. Picasso's twMerge is extendTailwindMerge from tailwind-merge (packages/picasso-tailwind-merge/src/twMerge.ts:35), which accepts the same input types as twJoin: strings, arrays, and falsy values (false, null, undefined, '') are filtered out.

So all of these work directly in twMerge(...):

condition && 'class' — Drawer.tsx:112, PaginationButton.tsx:20-22
condition ? 'a' : 'b' — Dropdown.tsx:271, DropdownArrow.tsx:21
Nested arrays — PageHeadBase.tsx:74 uses ['py-3', rightPadding && 'pr-8']
What's NOT supported is the clsx object syntax ({ active: isActive }). If you need that, you'd have to add clsx/cx — but the established Picasso pattern is the && / ternary forms above, which is what reviewers expect.

twJoin is also re-exported for the case where you just need to concatenate without conflict-resolution.

```

**Why both?** `cx` only joins strings conditionally. If a wrapper applies `px-4` and the consumer passes `px-2`, `cx` alone produces `"px-4 px-2"` — both classes ship to the DOM and Tailwind's last-wins rule becomes order-of-CSS-rules dependent and brittle. `twMerge` deduplicates Tailwind-conflicting classes deterministically: **rightmost class wins regardless of source order**.

**Do NOT introduce `clsx`** — Picasso already ships `classnames`. The `cn = clsx + tailwind-merge` helper common in external tutorials maps to `twMerge(cx(...))` here.

(source: `base-ui-styling.md §3.1`, `rules/styling.md §"Composition"`)

### Default classes + consumer override · **RULE**

The basic wrapper pattern: the kit owns its default classes; consumers override via standard `className`.

```tsx
export function CheckboxRoot({ className, ...props }: Checkbox.Root.Props) {
  return (
    <Checkbox.Root
      className={twMerge(

>         cx(

Reviewer: ```suggestion
```

          // layout
          'flex size-4 shrink-0 items-center justify-center rounded border',
          // colors
          'border-neutral-700 bg-white text-white',
          // state
          'data-[checked]:bg-neutral-900 data-[checked]:text-white',
          // focus
          'focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-neutral-900',
        ),
        // consumer override wins (rightmost in twMerge)
        className,
      )}
      {...props}
    />
  )
}
```

**Rules of the road**:

1. **Consumer `className` is always last** — rightmost in `twMerge(...)` wins.
2. **Never spread `...props` after explicit `className`** — the consumer's `className` is in `props` and silently overrides your merged value. Pass `className` explicitly, then spread the rest.
3. **Group your defaults by concern** (layout / colors / state / focus). Future-you will thank present-you.

(source: `base-ui-styling.md §3.2`)

### `className` as a function of state · **Preferred**

Every part exposing state accepts a function form:

```tsx
<Switch.Thumb
  className={(state) =>

>     twMerge(cx(

Reviewer: ```suggestion
    twMerge(
```

### Dynamic values · **Convention**

When a value really must be computed at runtime, use Tailwind's arbitrary-value syntax for discrete enums (purgeable) or `style` for true computed numbers:

```tsx
// Arbitrary value — purgeable when size is a finite enum
<div className={`w-[${size * 4}px]`} />

// style for numeric interpolation — Tailwind can't enumerate this
<div style={{ width: size * 4 }} />
```

(source: `rules/styling.md §"Dynamic values"`)

### Conditionals · **Convention**

Plain ternaries or Tailwind's data-attribute selectors:

```tsx
// Good
className={cx({ 'm-0': expanded, 'm-2': !expanded })}

// Good — data-attribute driven (lets parent styling participate)
<div data-state={expanded ? 'open' : 'closed'} className="data-[state=open]:bg-blue-500" />

// Bad — JSS parent-ref
'&$expanded': { margin: 0 }
```

(source: `rules/styling.md §"Conditionals"`)

### Hover / focus / disabled / responsive · **Convention**

Use Tailwind variant prefixes, not state-tracking JS:

```
hover:bg-blue-500
focus:ring-2 focus:ring-blue-400
disabled:opacity-50
md:flex lg:gap-12
```

(source: `rules/styling.md §"Hover / focus / disabled / responsive"`)

---

> ## 9. JSS → Tailwind translation (migration PRs)

Reviewer 2: Idk, this seems to me that it should be described as separate story on picasso like we had previously https://picasso.toptal.net/?path=/story/tutorials-styled-components--styled-components

To have single source of truth, who knows how it can change in the future. 

I understand that tw is optional, but styles components are optional as well and still they have their own picasso page.

Author: It's just for migration and when migration finish we will rework it for tailwind.

|---|---|---|
| `padding: spacing(1)`        | 8px   | `p-2`  |
| `padding: spacing(2)`        | 16px  | `p-4`  |
| `padding: spacing(3)`        | 24px  | `p-6`  |
| `padding: spacing(4)`        | 32px  | `p-8`  |
| `marginLeft: spacing(0.5)`   | 4px   | `ml-1` |
| `marginLeft: '1rem'`         | 16px  | `ml-4` |
| `marginRight: '0.5rem'`      | 8px   | `mr-2` |

### Color

MUI palette → Picasso tokens (most common):

| JSS                                 | Picasso Tailwind |
|---|---|
| `color: palette.text.primary`       | `text-graphite-800` |
| `color: palette.text.secondary`     | `text-graphite-700` |
| `color: palette.grey.dark`          | `text-gray-700` |
| `color: palette.primary.main`       | `text-blue-500` |
| `color: palette.primary.dark`       | `text-blue-600` |
| `color: palette.success.main`       | `text-green-500` |
| `color: palette.error.main`         | `text-red-500` |
| `color: palette.warning.main`       | `text-yellow-500` |
| `backgroundColor: palette.common.white` | `bg-white` |
| `backgroundColor: palette.grey.light`   | `bg-gray-100` |
| `backgroundColor: palette.background.paper` | `bg-white` |

> | `borderColor: palette.divider`      | `border-gray-300` |

Reviewer 2: My previous comment was specifically about this, since colors are the only things I can imagine can be changed. And they are picasso dependent, not common industry standards as for example spacings.

Author: Agree, duplication. It should point to /tokens/picasso-tailwind-tokens.md

  ref
) { /* ... */ })
```

> Documented variance: `Dropdown.tsx` exposes `defaultProps?: Partial<PropsWithBaseSpacing>` in its Props type for overload support — that's a type-level field name, not the runtime anti-pattern. Single-component carve-out; do NOT model new code on it.

(source: `code-standards.md §"Props interface declaration"`)

### JSDoc rules · **RULE** (24/28 have 100% public-prop coverage)

- **Required** on EVERY public Props interface field.
- Format: single-line `/** description */` immediately above the property.

```ts
interface Props extends BaseProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
}
```

- **Forbidden on passthrough internal props.** These surface in TS doc generation as public API. Reserved-props list (do NOT JSDoc, pass through silently):
  - `data-private?: string` — framework hook (theme access / analytics). Used internally by Tooltip, Slider, Switch, Checkbox.
  - `data-testid?: string` — comes from `BaseProps`; never declare per-component.

>   - `ownerState`, MUI-Base / `@base-ui/react` injected props — implementation detail of the underlying primitive.

Reviewer: I think this is too specific for `ownerState` here, for particular case

  If your migration touches a component that uses these, preserve the pass-through without adding JSDoc.
- For `@deprecated` JSDoc tags: should include a Jira reference `[ABC-1234]` or URL. ESLint `todo-plz/ticket-ref` is configured as **warn** (not error), so non-compliant `@deprecated` tags ship without blocking CI — reviewers consistently flag this regardless.

(source: `code-standards.md §"JSDoc rules"`)

### Import conventions · **RULE**

**Order** — enforced + auto-fixable by `import/order` (error) in `@toptal/davinci-syntax`'s ESLint config. Groups: `[builtin, external]` → `internal` → `[parent, sibling, index, unknown]`, with `newlines-between: always`. Run `pnpm davinci-syntax lint code packages/base/<NAME>/src` to auto-fix.

> **Barrel imports — RULE (NEVER deep paths)**: `import { StandardProps, SizeType } from '@toptal/picasso-shared'`, not `import { StandardProps } from '@toptal/picasso-shared/src/...'`. Survey: 28/28 conform.

Reviewer: nit: I think this is enforced by eslint autofixable rules

Author: Yeah, just cite rule.

(source: `code-standards.md §"*.example.tsx story files"` + `docs/contribution/creating-examples.md`)

---

## 4. Types & casting

Zero violations of these rules across the 28 migration-scope components today. Strict enforcement.

### No `any` in component source · **RULE** (ESLint `error`)

ESLint `@typescript-eslint/no-explicit-any` is **error** in source, off in tests. Tests may cast mock objects.

**Out-of-scope legacy**: 2 pre-existing `any` violations exist in untouched code (`TreeViewContainer.tsx:43`, `List/utils/generateListItems.tsx:6`) and 2 legacy `// @ts-ignore` comments in `Dropdown.tsx:223,227` with TODO+Jira markers. Those are explicit refactor markers, NOT precedents to extend.

(source: `code-standards.md §"Type-narrowing & casting"`)

### No `as unknown as T` blanket casts · **RULE**

In component source. Tests may cast mock objects.

(source: `code-standards.md §"Type-narrowing & casting"`)

### No bare `// @ts-ignore` · **RULE**

If you absolutely need to suppress a type, use `@ts-expect-error <reason>` so the suppression is self-documenting and fails the build when the underlying type drifts.

(source: `code-standards.md §"Type-narrowing & casting"`)

> ### The "prop-by-prop boundary" — canonical resolution for root-element-type mismatches · **RULE**

Reviewer: I think we should instruct to keep component interface untouched instead

Reviewer: those instructions look too specific

Author: We need to cover this case as strategy when base-ui <> picasso interface as way to keep picasso interface same. I will make it more condensed as principle with one example.

**To find which props to destructure**: open `node_modules/@base-ui/react/<group>/<part>/<Part>.d.ts` and diff its `*.Props` against your public `Props`. The NAME-OVERLAPS-WITH-DIFFERENT-TYPES intersection is your destructure list. Everything else is runtime-compatible and spreads safely. Under Picasso's `strict: true` tsconfig, event-handler element variance can still surface a `tsc` error on `...rest` even when the destructure list is correct — see "TS variance" below. For Tier 0 components, typically 1–3 props.

**Anti-patterns to avoid** (both forbidden):

- **Blanket cast** `as unknown as BaseUISwitch.Root.Props` — silences the type checker without addressing the mismatch.
- **Exhaustive allowlist** (manually picking 4–6 props to forward) — drops every other prop at runtime; reviewers call this the "typed but no-op" anti-pattern.

If you find yourself destructuring 6+ props, you're sliding into the exhaustive-allowlist anti-pattern — re-read the library's `.d.ts`. Cite this section + `practices.md §"API preservation"` directly in PR replies when reviewers raise the question.

**TS variance: when `tsc --strict` rejects `...rest` · sanctioned escape**

> Picasso's `tsconfig.base.json` has `"strict": true` → `strictFunctionTypes` is on. If your `Props` extends an element-specific HTML attributes type (e.g. `ButtonHTMLAttributes<HTMLButtonElement>`) and the base-ui part renders a different element (Switch → `<span>`), `tsc` will reject `{...rest}` even when the destructure list is correct — because `MouseEventHandler<HTMLButtonElement>` is not assignable to base-ui's span-typed `BaseUIEvent<MouseEvent<HTMLSpanElement>>` due to function-parameter contravariance.

Reviewer: I think this is not needed, also too specific rules

This is a real, runtime-safe variance: React synthetic event handlers fire identically regardless of `currentTarget` element type. Do NOT narrow the public `Props` to fix this — reviewers explicitly reject contract narrowing (PR #4965 review 2026-05-20 16:10). Express the bridge once at the boundary, NOT in JSX:

```tsx
const rootRest = rest as Omit<
  BaseUISwitch.Root.Props,
  'checked' | 'disabled' | 'id' | 'value' | 'className' | 'style' | 'onCheckedChange'
>

return <BaseUISwitch.Root {...rootRest} … />
```

This is sanctioned, not a blanket cast: plain `as` (not `as unknown as`) between types with structural overlap; `Omit`s the props you explicitly override; local typed binding (not JSX-site); leave a one-line comment citing the variance reason. Reference: PR #4965 iter 11 (2026-05-28).

(source: `code-standards.md §"prop-by-prop boundary"` + `§"TS variance: when tsc --strict rejects ...rest"`, `practices.md §"API preservation"`)

> ### Type alignment at the boundary · **RULE**

Reviewer: no castings 😃

Author: It's not a rule to cast, but if it's needed to do it in specific way.

Cast at the type BOUNDARY (helper return type, local typed binding), NOT at the JSX call site:

```tsx
// Preferred — hoist cast into helper return type:
const getClickHandler = (
  loading?: boolean,
  handler?: Props['onClick']
): BaseUIButton.Props['onClick'] =>
  (loading ? noop : handler) as BaseUIButton.Props['onClick']

// Then JSX is clean:
<BaseUIButton onClick={getClickHandler(loading, onClick)} />
```

This isolates the type-system compromise to one location instead of sprinkling casts across the render. Reviewers can verify the boundary cast in one place.

(source: `code-standards.md §"prop-by-prop boundary"`, `rules/base-ui-react-api-crib.md §"Type alignment at the boundary"`)

### Polymorphic + ref forwarding · **RULE**

- `forwardRef<HTMLButtonElement, Props>(...)` already types `ref` correctly. Don't cast `ref` at the JSX site.

> - **Drop Picasso-only props from `rest` before spreading to a `@base-ui/react` part.** When `rest` includes Picasso-specific props the underlying part doesn't accept, destructure them out before the spread — don't paper over the mismatch with `as BaseUIButton.Props`. NEVER fall back to `any`.

Reviewer: probably this is not necessary and 

> - `forwardRef<HTMLButtonElement, Props>(...)` already types `ref` correctly. Don't cast `ref` at the JSX site.

we can have as a separate rule

| **5. `style` prop (static or function-of-state)** | Inline styles | Last resort — computed values that cannot be expressed as classes |

Every part exposes the consistent signature:

```ts
className?: string | ((state: State) => string | undefined);
style?:     React.CSSProperties | ((state: State) => React.CSSProperties | undefined);
render?:    ReactElement | ((props: HTMLProps, state: State) => ReactElement);
```

(source: `base-ui-styling.md §2`)

### `className` composition — `twMerge(...)` (with optional `cx`) · **RULE**

Every Tailwind-on-headless codebase needs a class-merging pipeline that resolves conflicts. Picasso uses `twMerge` from `@toptal/picasso-tailwind-merge`. `cx` from `classnames` is **optional** — only needed for clsx-object-syntax.

**Default form — `twMerge(...)` alone:**

```ts
import { twMerge } from '@toptal/picasso-tailwind-merge'

twMerge(
  'px-4 text-sm',
  isLarge && 'px-6 text-base',         // conditional via && — works directly
  variant === 'primary' ? 'bg-blue-500' : 'bg-transparent',
  className                            // consumer override LAST
)

> ```

Reviewer: it's a bit strange rule 🤔 I understand it came from the previous mistake, but still, I'd expect agent to figure out that on it's own. Maybe max giving some component code example

Author: Yeah it came from learning, maybe it was done when I had been using sonnet model. I would keep it.

Example of the legacy shape (only valid for Tier 3.b's continued v0-API surface):

```tsx
// LEGACY — Tier 3.b only (OutlinedInput.tsx:174-202)
<Input
  slots={{ input: CustomInput }}
  slotProps={{
    root: { className: twMerge('...', className) },
    input: { className: twMerge('...', inputClassName) },
  }}
/>
```

**End-state**: Tier 3.b consolidates to v1 per-part styling once consumers migrate (see Appendix A.2). Anything else that survives a migration with `slotProps` is unmigrated — flag for the next iter.

(source: `practices.md §"Slot-based styling — LEGACY Tier 3.b ONLY"`, `base-ui-styling.md §Appendix·1`)

### Input-bearing slots · **Convention**

For input-bearing slots (`Slider.Thumb`, `Switch.Input`): hide visible native `<input>` via:

```
[&_input]:!top-auto [&_input]:!left-auto
[&_input]:![clip-path:none] [&_input]:[clip:rect(0,0,0,0)]

> ```

Reviewer: I think this is not needed

### Async focus management · **Convention**

Base UI's rAF-deferred `FloatingFocusManager` diverges from `@mui/base`'s synchronous focus. Add a `useIsomorphicLayoutEffect` blur-on-open shim with an explanatory comment when reviewers flag visual-snapshot regressions tied to focus timing.

(source: `practices.md §"@base-ui/react idioms"`)

### `disablePortal` emulation · **Convention**

Conditionally omit `<Drawer.Portal>` wrapper rather than searching for a non-existent prop. Some `@mui/base` portal/behavior props (like `disablePortal` on Drawer) have no direct prop equivalent in `@base-ui/react/drawer`, but can be emulated by conditionally omitting the wrapper.

(source: `practices.md §"@base-ui/react idioms"`)

### Responsive spacing utilities · **Convention**

Components accepting breakpoint-aware spacing (e.g., Dropdown's `offset?.top` / `offset?.bottom`) should use `makeResponsiveSpacingProps()` from `@toptal/picasso-provider` to generate responsive Tailwind classes dynamically. See `Dropdown.tsx:106-109,236-242` for the canonical usage. Do NOT hand-roll responsive class strings for spacing values that should match the breakpoint API.

> (source: `practices.md §"@base-ui/react idioms"`)

Reviewer: I think this is not needed and agent will figure this out on it's own if it has access to docs

---

## 6. Tailwind class composition

### `styles.ts` as pure functions returning `string[]` · **RULE** (established by Button canonical)

Class-building logic lives in `styles.ts` as **pure functions returning `string[]`** (Button pattern, 14/28 conform; 8/28 use `cx` inline).

```ts
// styles.ts
export function createSizeClassNames(size: 'small' | 'medium' | 'large'): string[] {
  switch (size) {
    case 'small':  return ['text-button-small',  'px-3', 'h-8']
    case 'medium': return ['text-button-medium', 'px-4', 'h-10']
    case 'large':  return ['text-button-large',  'px-6', 'h-12']
  }
}
```

(source: `code-standards.md §"Tailwind class composition"`, `rules/styling.md §"Helper-fn shape"`)

> ### `twMerge` with caller's `className` LAST · **RULE**

Reviewer: it might be not always correct, also, I think code examples are not needed for this one

Author: It's happening in 50% of components, I can change it to preferred and remove example.

Merge in `Component.tsx` via `twMerge(coreClassNames, variantClassNames, ..., className)` — **user-supplied `className` LAST** so consumer overrides win.

The wrong order (`twMerge(className, structural)`) silently breaks consumer customization (Drawer iter 3 lesson).

```tsx
// Good — consumer wins
className={twMerge(cx(
  'inline-flex items-center justify-center',
  ...createSizeClassNames(size),
  ...createVariantClassNames(variant),
), className)}

// Bad — structural classes silently override the consumer's className
className={twMerge(className, cx(
  'inline-flex items-center justify-center',
  ...createSizeClassNames(size),
))}
```

(source: `practices.md §"Tailwind & class composition"`, `base-ui-styling.md §3.2`)

> ### `cx` for conditional grouping; `twMerge` for conflict resolution · **RULE**

Reviewer: I think cx can be just never used

- `cx` for conditional/boolean grouping when no conflict resolution is needed.
- `twMerge` for merging conflicting Tailwind classes (one wins via the merge algorithm).

```tsx
// cx — conditional groupings without conflicts
className={cx('p-4', loading && 'opacity-50')}

// twMerge wrapping cx — when caller-supplied className may conflict
className={twMerge(cx('p-4', loading && 'opacity-50'), className)}
```

(source: `code-standards.md §"Tailwind class composition"`)

> ### `@toptal/picasso-tailwind-merge` extensions · **Convention**

Reviewer: this is not needed I think

Author: Yeah, maybe just pointer.

`@toptal/picasso-tailwind-merge` wraps `tailwind-merge` with Picasso-specific extensions — see `packages/picasso-tailwind-merge/src/twMerge.ts`:

- Custom font sizes: `text-2xs`, `text-xxs`, `text-button-{small|medium|large}`, `font-inherit-size`
- Custom weights: `font-regular`, `font-semibold`, `font-inherit-weight`
- Text-align preservers: `text-align-inherit`, `text-start`, `text-end`

Always import `twMerge` from `@toptal/picasso-tailwind-merge` — never the upstream `tailwind-merge`.

(source: `code-standards.md §"Tailwind class composition"`)

### Tier 0 light-path: Tailwind class composition stays as-is · **RULE** (migration-period)

> Don't rewrite styles when the package swap is the only change. The Tailwind groundwork was done in the `@mui/base` era — the win of that migration was leaving Tier 0 components on `cx`/`twMerge` already.

Reviewer: this does not make sense to me

Author: It's migration internal thing to optimize migration but yes it can be removed or simplified.

(source: `practices.md §"Tailwind & class composition"`)

### Tier 1+ heavy-path: read `rules/jss-to-tailwind-crib.md` IN FULL · **RULE** (migration-period)

Before touching any JSS. The cribsheet's worked examples cover parent-ref selectors, dynamic class-from-state, raw hex → tokens, pseudo selectors, and `theme.spacing` → gap utilities. Full table inlined in §9 below.

(source: `practices.md §"Tailwind & class composition"`)

### Twmerge boundary

When you accept a `className` prop from a consumer, merge with `twMerge` so consumer-provided utilities can override component defaults:

```tsx
import { twMerge } from '@toptal/picasso-tailwind-merge'

<div className={twMerge(cx('p-4 bg-blue-100', baseClasses), className)} />
```

> The consumer should always be able to win.

Reviewer: I think should be removed

Author: Yes, it should just be simple className last rule.

(source: `rules/styling.md §"Twmerge boundary"`)

---

## 7. CSS specificity ladder (no `!important`)

> **Restructured 2026-05-26 (commit `3e9766e8`)**: the override-preference ladder + narrowed inline-vs-inline CSS specificity hierarchy are now TWO separate concepts. This section now reflects both.

There are two related but separate ladders. Reviewers enforce both:

- **§7.1 Override-preference ladder** (`base-ui-styling.md §7.1`): which mechanism to reach for FIRST when you need to override Base UI defaults. Conceptual — start at the cheapest rung.
- **§7.2 CSS specificity hierarchy** (`code-standards.md §"CSS specificity hierarchy"`): narrow case of what wins via the CSS cascade when both consumer and Base UI emit inline `style`. Mechanistic — describes what beats what.

> Do NOT conflate them. The preference ladder is about *which tool to pick*; the specificity hierarchy is about *what wins in the cascade*.

Reviewer: I think this can be removed, only important info here is `no !important`

Author: Yes, it could be simplified in a way of `no !important` > data-attr > render > last inline style

### 7.1 Override-preference ladder · **RULE**

Pick the lowest rung (lowest = cheapest). Reviewers will block PRs that skip rungs.

| Rung | Mechanism | When to use |
| --- | --- | --- |
| **-1** | **Don't override** | **Check this FIRST.** Restructure DOM to use Base UI's native part hierarchy, OR embrace Base UI's geometry and accept any sub-pixel diff as approved improvement. The override that doesn't exist is the cheapest one to maintain. |
| 1 | `className` + `data-[…]:` variants | State-driven styling that has a documented data attribute (e.g., `data-[checked]:bg-blue-500`). |
| 2 | `className` function-of-state form | State-driven styling where the value depends on multiple state combinations (rare). |
| 3 | `render` prop with `mergeProps` filtering | Change rendered element type, compose with custom components, OR strip/filter Base UI's injected `style` (e.g., remove `translate` while keeping `left: X%`). |
| 4 | `useRender` + `mergeProps` | Build your own primitive that exposes its own `render` prop. |
| 5 | inline `style` on the part | Genuine last resort: when render-prop filtering is impractical, AND the value can't be expressed as a class, AND rung -1 doesn't apply. |

> | ⊘ | `!important` | NEVER. If you're here, you skipped a rung. |

Reviewer: this is already covered above, I think can be removed

Author: I like to have this defined as simple table, easier to unerstand than bullets.

3. **Trigger**: if you're adding inline `style` or `!important` to match a legacy baseline byte-for-byte, you're at rung 5 or worse — step back to rung -1.

#### Slider Thumb worked example

```tsx
// Rung 3 — render-prop filtering (preferred over rung 5)
<BaseUISlider.Thumb
  render={(props) => {
    const { translate, ...keepStyle } = props.style || {}
    return <span {...props} style={keepStyle} className={thumbClassName} />
  }}
/>

// Rung 5 — consumer-added inline style (only if rung 3 is impractical)
<BaseUISlider.Thumb className={thumbClassName} style={{ translate: 'none' }} />

// Rung -1 — best: remove the legacy `-mt -ml` margins that the override was defending;
// accept Base UI's `translate: -50% -50%` centering; classify ~1.5px Happo diff as "intentional improvement"
<BaseUISlider.Thumb className={thumbClassName} />   // -mt/-ml removed from thumbClassName
```

> (source: `base-ui-styling.md §7.1`)

Reviewer: Too specific, I think we can remove that

```tsx
<Slider.Thumb style={{ translate: 'none' }} ... />  // YES — wins via mergeProps right-wins
<Slider.Thumb className='![translate:none]' ... />   // NO — !important is forbidden
```

**Why inline `style` exists at all**: when @base-ui/react renders, props are merged into the final element via `useRenderElement`. `style` is merged last with `mergeObjects(componentInternalStyle, consumerStyle)` — rightmost (yours) wins per property. This is the headless-kit's contract; it's designed for you to override its defaults *when you need to*. But before reaching for rung 5, check whether §7.1 rung -1 (don't override) or rung 3 (render-prop filtering) applies.

**Limits**: covers per-property overrides (e.g. `translate`, `position`). Does NOT cover style based on internal state (`data-focused`, `data-orientation`) — those go through `data-[…]:` per §7.1 rung 1.

> (source: `code-standards.md §"CSS specificity hierarchy for overriding @base-ui/react internal inline styles"`)

Reviewer: the same, I think it's already covered above, so no need to reiterate with more specific rules here

> ### ANTI-PATTERN — imperative `ref` callbacks mutating `.style` · **RULE — FORBIDDEN, no exceptions**

Reviewer: makes sense, but I think we already explained in CSS overrides section how to DO that, so we don't need how to NOT DO here

Author: I know, but I would like to keep this as additional gate.

Examples that violate this:
- `inputRef={node => { node.style.margin = '0' }}`
- `ref={n => n?.style.setProperty('translate', 'none')}`
- `useCallback` wrapping any `.style.X = …` assignment passed as a slot ref.

**Why this is forbidden**: it bypasses CSS, breaks responsive style changes, isn't tree-shaken, and creates a runtime side-channel that fights the design system. Use the §7.1 preference ladder instead.

Earlier Switch migration code (iter 2) used this pattern; treat any such occurrence as a defect to remove during cleanup, NOT a precedent to extend. This rule has no "one-off compromise" carve-out.

> #### Explicitly rejected justifications (do not cite these to defend the pattern)

Reviewer: I think too specific instructions and need to be removed

- **"base-ui inline `style=""` can't be overridden by CSS"** → true that only inline `style` (§7.1 rung 5) beats inline `style` via CSS specificity. The idiomatic fix is the `style` prop on the part (mergeProps merges consumer style with rightmost-wins) — NOT an imperative ref-mutation. And before reaching for rung 5, check whether rung -1 applies.
- **"Cited as a precedent in practices.md / lessons-learned"** → no, it's an anti-pattern. Older wording that framed it as a "compromise" was contamination superseded by this rule.

Reviewers consistently rejected this pattern across Switch iters 2/3/9. Treat any new instance as a defect, not a precedent.

**Imperative `ref` callbacks remain valid for non-style concerns** (focus management, measurement, third-party library handles, port resize observers).

(source: `code-standards.md §"CSS specificity hierarchy ..."`, `practices.md §"API preservation"`)

---

## 8. Tokens & forbidden CSS patterns

### Picasso Tailwind tokens · **RULE**

Use Picasso tokens by their semantic name where possible:

```

> Good: text-graphite-800, bg-blue-100, shadow-2 (modal), p-4 (16px)

Reviewer: ```suggestion
Good: text-graphite-800, bg-blue-100, shadow-2, p-4
```

Bad:  text-[#262D3D], bg-[#EDF1FD], shadow-[0_4px_8px_0_rgba(0,0,0,0.08)], p-[16px]
```

If you find yourself reaching for an arbitrary value, double-check `tokens/picasso-tailwind-tokens.md` first.

(source: `rules/styling.md §"Token usage"`)

### `TODO(tokens):` when no canonical token exists · **RULE**

When no canonical token exists for a CSS value, keep the literal + add a `// TODO(tokens): <description>` comment. Don't invent a new token; that's a coordinated design-system change.

```tsx
'bg-[#4269D6]',     // TODO(tokens): brand-blue-variant — designer can confirm canonical name
'rounded-[6px]',    // TODO(tokens): 6px isn't on the 4px scale; verify if intentional
```

> These comments surface in the P1-FIG-03 audit so designers can resolve later.

Reviewer: I think this should be removed, we shouldn't advice using non-token variables at all

Author: I would keep it as last resort so we can act and discuss with designers.

(source: `rules/styling.md §"Token usage"`, `practices.md §"Tailwind & class composition"`)

### No CSS files · **RULE**

No `.css`, `.scss`, `.module.css`. Anything CSS-shaped lives in Tailwind classes or helper-returned arrays.

> (source: `rules/styling.md §"What to avoid"`)

Reviewer: I think should be changed to a one-liner in ./AGENTS.md file

```
- use Tailwind CSS for styling
```

### No JSS objects · **RULE**

No `makeStyles`, `createStyles`, `withStyles`. No `&$selector` parent-refs.

For variant-driven classes, return a `string[]` from a pure function in `styles.ts` (Button precedent) and merge with `twMerge`. Do NOT introduce new JSS `classes` maps. Existing JSS-using components (Tier 2 Radio, Tier 3 Page, sibling-package picasso-charts/RTE) are migration targets, not pattern sources.

`docs/contribution/css-naming.md` describes MUI v4 + JSS conventions (`root` + `rootFull`/`rootShrink` for variants, `cx({ [classes.active]: active })`, etc.). **These are PRE-migration patterns — LEGACY, do not follow.**

> (source: `rules/styling.md §"What to avoid"`, `practices.md §"css-naming.md is LEGACY"`)

Reviewer: I think same here, we should not even mention this, it will work correctly by default, I think, considering `use Tailwind` instructions

Author: Ok, but we need to be care first of removing it after migration. Once we finish migration we can clean rules and files.

### No `style={{...}}` for static values · **RULE**

Only use inline `style` when the value is computed at runtime from props. Numeric interpolation is OK; static styles must move to Tailwind.

```tsx
// Bad — static color in inline style
<div style={{ color: 'red' }} />

// Good — Tailwind class
<div className="text-red-500" />

// Acceptable — value computed from props
<div style={{ width: size * 4 }} />
```

> (source: `rules/styling.md §"What to avoid"`, `base-ui-styling.md §3.5`)

Reviewer: I think not needed rule

### Dynamic values · **Convention**

When a value really must be computed at runtime, use Tailwind's arbitrary-value syntax for discrete enums (purgeable) or `style` for true computed numbers:

```tsx
// Arbitrary value — purgeable when size is a finite enum
<div className={`w-[${size * 4}px]`} />

// style for numeric interpolation — Tailwind can't enumerate this
<div style={{ width: size * 4 }} />
```

> (source: `rules/styling.md §"Dynamic values"`)

Reviewer: that's a very niche thing, maybe we can specify particularly how size prop should work

### Conditionals · **Convention**

Plain ternaries or Tailwind's data-attribute selectors:

```tsx
// Good

> className={cx({ 'm-0': expanded, 'm-2': !expanded })}

Reviewer: ```suggestion
className={twMerge({ 'm-0': expanded, 'm-2': !expanded })}
```

Author: Are you sure twMerge supports `clsx` syntax, I know it support `cx`?

Author: Only this `className={twMerge(expanded ? 'm-0' : 'm-2')}`

// Good — data-attribute driven (lets parent styling participate)
<div data-state={expanded ? 'open' : 'closed'} className="data-[state=open]:bg-blue-500" />

// Bad — JSS parent-ref
'&$expanded': { margin: 0 }
```

(source: `rules/styling.md §"Conditionals"`)

### Hover / focus / disabled / responsive · **Convention**

> Use Tailwind variant prefixes, not state-tracking JS:

Reviewer: ```suggestion
Use Tailwind variant prefixes:
```
